models/turbine_models/custom_models/sdxl_inference/unet.py

```diff
@@ -67,12 +74,85 @@ def forward(
             added_cond_kwargs=added_cond_kwargs,
             return_dict=False,
         )[0]
+        if self.do_classifier_free_guidance:
+            noise_preds = noise_pred.chunk(2)
+            noise_pred = noise_preds[0] + guidance_scale * (
+                noise_preds[1] - noise_preds[0]
+            )
         return noise_pred
 
 
+def get_punet_model(hf_model_name, external_weight_path, precision="i8"):
+    from sharktank.models.punet.model import (
+        Unet2DConditionModel as sharktank_unet2d,
+        ClassifierFreeGuidanceUnetModel as sharktank_CFGPunetModel,
+    )
+    from sharktank.utils import cli
+
+    if precision == "i8":
+        repo_id = "amd-shark/sdxl-quant-models"
+        subfolder = "unet/int8"
+        revision = "82e06d6ea22ac78102a9aded69e8ddfb9fa4ae37"
+    elif precision in ["fp16", "fp32"]:
+        repo_id = hf_model_name
+        subfolder = "unet"
+        revision = "76d28af79639c28a79fa5c6c6468febd3490a37e"
+
+    def download(filename):
+        return hf_hub_download(
+            repo_id=repo_id, subfolder=subfolder, filename=filename, revision=revision
+        )
+
+    results = {
+        "config.json": download("config.json"),
+        "params.safetensors": download("params.safetensors"),
+    }
+    if precision == "i8":
+        results["quant_params.json"] = download("quant_params.json")
+        output_path = external_weight_path.split("unet")[0] + "punet_dataset_i8.irpa"
+        ds = get_punet_dataset(
+            results["config.json"],
+            results["params.safetensors"],
+            output_path,
+            results["quant_params.json"],
+            base_params=None,
+        )
+    else:
+        ds = get_punet_dataset(
+            results["config.json"],
+            results["params.safetensors"],
+            output_path,
+            base_params=None,
+        )
+
+    cond_unet = sharktank_unet2d.from_dataset(ds)
+    mdl = sharktank_CFGPunetModel(cond_unet)
+    return mdl
+
+
+def get_punet_dataset(
+    config_json_path,
+    params_path,
+    output_path="./punet_dataset_i8.irpa",
+    quant_params_path=None,
+    quant_params_struct=None,
+    base_params=None,
+):
+    from sharktank.models.punet.tools import import_brevitas_dataset
+
+    import_brevitas_dataset.main(
+        [
+            f"--config-json={config_json_path}",
+            f"--params={params_path}",
+            f"--quant-params={quant_params_path}",
```

Comment:
I don't think this will work for the fp16 version like this (the literal value "None" will try to load a file named "None").

There is an importer right next to this for a stock huggingface dataset which was intended to be used for this case. Should probably switch here, but judging from above, I'm guessing that fp16 punet has not been tested yet. Can flush when it is.